helix-term/src/commands.rs

```diff
@@ -1057,11 +1062,81 @@ fn goto_file_impl(cx: &mut Context, action: Action) {
                 .to_string(),
         );
     }
+
+    // Clippy output format.
+    let regex_file_row_col = Regex::new("(?P<path>.[^:]+):(?P<row>\\d+):(?P<col>\\d+)").unwrap();
```

Comment:
This is too specific to Rust, and so we probably want to use a more generic way of detecting this information. Ideally, I'd imagine that you could do something like `cargo clippy | regex_query | hx`.

Reply:
I think ideally we should support all common formats so that Helix always works out of the box.

In fact, I'd argue `PATH:ROW:COL` and `PATH:ROW` are not very Rust specific. A lot of compilers print paths in this way. For example, webpack prints in this exact format as well.

Also, one more argument for not relying on terminal piping has been specified in the linked issue (#5259): this feature would be very useful once integrated terminal is implemented, where you can't do any text transformation at all, so this workaround doesn't apply anymore.

Reply:
It can also be turned into a lazy static.

Reply:
What about ``PATH COL:ROW``? This vscode issue gives some examples:  https://github.com/microsoft/vscode/issues/140780

Reply:
> It can also be turned into a lazy static.

Fixed. Thanks for pointing that out. I wasn't aware Helix already had lazy static set up.

Reply:
> What about `PATH COL:ROW`?

What do the maintainers think about this? @kirawi

To support this we simply have to allow the first `:` to be one of `:` and ` `, a trivial change in the regex.

Reply:
https://github.com/microsoft/vscode/blob/1f00f5a6a62b5c860333be498b9f314b93f8afcb/src/vs/workbench/contrib/terminal/browser/links/terminalLocalLinkDetector.ts#L57

You can find a list of all formats supported by vscode (including the corresponding regex) above.

Reply:
Ah this is helpful. Lemme change to use this list of regex instead. Thanks for the pointer!

Reply:
The vscode detector is so comprehensive that I think we can drop the existing trimming as well. I will update this PR by porting the whole thing over :))

I think with this powerful detector we can even change the `goto_file` feature from the current:

"take this selection (or WORD) as file path as is except some funky chars to be trimmed"

to

- if it's a single-char selection (aka. no selection), "take the whole line and try to find a file link which contains the current cursor position and open it";
- otherwise, "try to detect a file link _within_ the selection and open it".

What do you guys think?

Reply:
That would be such a cool feature, personally I'd love to have your suggestion as described

Reply:
I don'y think you need to consider the second case. Seems pretty unintuitive that the goto command wouldn't work if you have selected something within the link. Other commands like goto_definition only take the cursor into account so.it seems more consistent to just use the first approach you mentioned

Reply:
@pascalkuthe You're right. I proposed that because the current implementation distinguishing between the two cases. But now that I took a step back and only consider UX, doing that makes little sense. The special handling currently in place is likely to make up for the cases where auto-trimming etc. fails, which shouldn't be needed with the vscode detector port.

Reply:
I started looking into the vscode impl. The regexs make heave use of backreferences which are not supported in the `regex` crate. We likely need `fancy-regex` here.

I looked into the current deps tree and `fancy-regex` is not used. Porting the detector would therefore involve introducing new dependencies. Maybe @archseer can kindly weigh in on this? Thanks!

Reply:
It looks like a `splitn(3) 
` to me, we already have this parsing implemented somewhere in app init since we parse the file argument this way

Reply:
https://github.com/helix-editor/helix/blob/df1830ef28a7cb49abe31a18e4bd1bcfc7eb802a/helix-term/src/args.rs#L112-L121

Reply:
@archseer Ah I was actually referring to the file path detector used in vscode pointed out by @pascalkuthe:

https://github.com/microsoft/vscode/blob/1f00f5a6a62b5c860333be498b9f314b93f8afcb/src/vs/workbench/contrib/terminal/browser/links/terminalLocalLinkDetector.ts#L57

The simple case of `PATH:ROW:COL` already works pretty well in this PR with no additional dependency needed. The`fancy-regex` crate is needed to port the whole vscode detector over (which allows us to have to same file linking capability as vscode, which works pretty awesome imo).

Reply:
Sorry. I studied the vscode detector regexs deeper and it turns out that those are not meant to be backreferences. `\0` there actually intends to mean the unicode char `\u0000` if I'm not mistaken. In that case, we actually don't need `fancy-regex`. The existing `regex` crate should work just fine.

I'm kinda confused tho. In the JavaScript regex engine, `\0` should actually mean backtraces. I just checked in node and it seems to actually be the case:

```console
$ node
> new RegExp("\\0").exec("a")
null
> new RegExp("\\0").exec("\u0000")
[ '\x00', index: 0, input: '\x00', groups: undefined ]
```

So maybe, just maybe, the vscode detector has a bug here. They incorrectly used `\0` to mean the null terminator whereas `\0` actually means a backreference, and this bug was left undiscovered due to the rare occurrence of `\u0000`.

Reply:
Yeah I don't think you need to backtrack for the so.ole cases. I wanted to take a look at the regexes to confirm but didn't get the chance. I don't think we want to lull in fancy regex

Reply:
@pascalkuthe Yeah I was also wondering why they even needed backreferences at all. I guess I just trusted regex101 too much :/